[SPARK-14393][SQL] values generated by non-deterministic functions shouldn't change after coalesce or union

mengxr opened the pull request:
## What changes were proposed in this pull request?

When a user appended a column using a "nondeterministic" function to a DataFrame, e.g., `rand`, `randn`, and `monotonically_increasing_id`, the expected semantic is the following:
- The value in each row should remain unchanged, as if we materialize the column immediately, regardless of later DataFrame operations.

However, since we use `TaskContext.getPartitionId` to get the partition index from the current thread, the values from nondeterministic columns might change if we call `union` or `coalesce` after. `TaskContext.getPartitionId` returns the partition index of the current Spark task, which might not be the corresponding partition index of the DataFrame where we defined the column.

See the unit tests below or JIRA for examples.

This PR uses the partition index from `RDD.mapPartitionWithIndex` instead of `TaskContext` and fixes the partition initialization logic in whole-stage codegen, normal codegen, and codegen fallback. `initializeStatesForPartition(partitionIndex: Int)` was added to `Projection`, `Nondeterministic`, and `Predicate` (codegen) and initialized right after object creation in `mapPartitionWithIndex`. `newPredicate` now returns a `Predicate` instance rather than a function for proper initialization.
## How was this patch tested?

Unit tests. (Actually I'm not very confident that this PR fixed all issues without introducing new ones ...)

cc: @rxin @davies

mengxr committed:
add test cases

mengxr committed:
fix WholeStageCodegen

mengxr committed:
add initializeStateForPartition to Projection

mengxr committed:
add RDD.mapPartitionsWithIndexInternal

mengxr committed:
fix issue without whole stage codegen

mengxr committed:
Nondeterministic.setInitialValues => initializeStatesForPartition(partitionIndex)

mengxr committed:
test all code paths

mengxr committed:
fixed codegen fallback case

mengxr committed:
also initialize predicate

mengxr committed:
fix other nondeterministic functions

mengxr committed:
test all nondeterministic functions

mengxr committed:
fix partition initialization in local relation

mengxr committed:
minor

SparkQA commented:
**[Test build #67257 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67257/consoleFull)** for PR 15567 at commit [`4dc0255`](https://github.com/apache/spark/commit/4dc025593fb06e501b148c2625c0b2e5d72655bc).

SparkQA commented:
**[Test build #67258 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67258/consoleFull)** for PR 15567 at commit [`5b915a7`](https://github.com/apache/spark/commit/5b915a7c0e4f708f469baecce1d0cf6b472f0b07).

SparkQA commented:
**[Test build #67258 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67258/consoleFull)** for PR 15567 at commit [`5b915a7`](https://github.com/apache/spark/commit/5b915a7c0e4f708f469baecce1d0cf6b472f0b07).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67258/
Test FAILed.

SparkQA commented:
**[Test build #67260 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67260/consoleFull)** for PR 15567 at commit [`e2ebd88`](https://github.com/apache/spark/commit/e2ebd88b796c560bfa5653491956bdeaa7021c86).

SparkQA commented:
**[Test build #67257 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67257/consoleFull)** for PR 15567 at commit [`4dc0255`](https://github.com/apache/spark/commit/4dc025593fb06e501b148c2625c0b2e5d72655bc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67257/
Test FAILed.

SparkQA commented:
**[Test build #67260 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67260/consoleFull)** for PR 15567 at commit [`e2ebd88`](https://github.com/apache/spark/commit/e2ebd88b796c560bfa5653491956bdeaa7021c86).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67260/
Test FAILed.

mengxr committed:
fix catalyst expression tests

mengxr committed:
fix generated predicate

SparkQA commented:
**[Test build #67270 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67270/consoleFull)** for PR 15567 at commit [`6659795`](https://github.com/apache/spark/commit/6659795fac07e5046af04407454b235c012b9ef6).

rxin commented on core/src/main/scala/org/apache/spark/rdd/RDD.scala:
```diff
@@ -787,14 +787,26 @@ abstract class RDD[T: ClassTag](
+  /**
+   * [performance] Spark's internal mapPartitions method that skips closure cleaning.
+   */
   private[spark] def mapPartitionsInternal[U: ClassTag](
```
can we get rid of this?

mengxr replied:
There are 20+ probably valid use of `mapPartitionsInternal`. The main problem is that changing it to `mapPartitionsWithIndexInternal` doesn't really force people to initialize the partition.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:
```diff
@@ -274,12 +274,12 @@ trait Nondeterministic extends Expression {
 
-  final def setInitialValues(): Unit = {
-    initInternal()
+  final def initializeStatesForPartition(partitionIndex: Int): Unit = {
```
while you are at it, it'd be great to add some comments documenting the function ...

mengxr replied:
I don't want to overload the name `initialize`, which is a little vague, how about `initStates`? Again, the issue is that even with comments we cannot force users to initialize it.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:
```diff
@@ -185,6 +185,20 @@ class CodegenContext {
+   */
+  val partitionInitializationStatements: mutable.ArrayBuffer[String] = mutable.ArrayBuffer.empty
+
+  def addPartitionInitializationStatement(statement: String): Unit = {
```
any reason you are creating this rather than just using addMutableState?

mengxr replied:
I'm a little worried about introducing more issues by moving `initMutableStates` out from the constructor. The current implementation at least maintains the existing behavior if we missed `initializeStatesForPartition` somewhere.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:
```diff
@@ -274,12 +274,12 @@ trait Nondeterministic extends Expression {
 
   private[this] var initialized = false
```
should we change this to transient? then it will always get reset to false on a new partition.

mengxr replied:
will do

mengxr committed:
checked all mapPartitionsInternal usage

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SparkPartitionID.scala:
```diff
@@ -17,16 +17,15 @@
  */
 @ExpressionDescription(
-  usage = "_FUNC_() - Returns the current partition id of the Spark task",
+  usage = "_FUNC_() - Returns the current partition id",
```
hmmm this is behavior changing, and there is some value to the old partition id ...

rxin replied:
I'd consider introducing a new expression for the proper id and leave the old one as is.

mengxr replied:
I thought about this. But I don't think the current behavior is the expected behavior from users. This is the same issue as in `monotonically_increasing_id`.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GeneratePredicate.scala:
```diff
@@ -25,19 +25,26 @@ import org.apache.spark.sql.catalyst.expressions._
+   * This is used by non-deterministic expressions to set initial states.
+   * The default implementation does nothing.
+   */
+  def initializeStatesForPartition(partitionIndex: Int): Unit = {}
```
to make this safer, i'd create an internal variable "isInitialized" similar to the one in nondeterministic expression, and assert in eval if isInitialized is false
.

mengxr replied:
I didn't test. Would doing that hurt the performance?

rxin replied:
I don't think so, since it is in the interpreted path which is already very slow. Also in the normal case the condition will always be false, so CPU branch prediction should work its magic.

rxin commented:
Reviewing this code makes me realize how painful it is when project/filter are just scala functions ... it'd be much easier to review if they have methods defined (e.g. eval, or execute) ...

rxin commented:
So my biggest question is whether you've changed all the places to call initialize where projection/predicate are used.

SparkQA commented:
**[Test build #67277 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67277/consoleFull)** for PR 15567 at commit [`80f26c6`](https://github.com/apache/spark/commit/80f26c6e010b18e712e9514146975f21af9ac8ab).

SparkQA commented:
**[Test build #67270 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67270/consoleFull)** for PR 15567 at commit [`6659795`](https://github.com/apache/spark/commit/6659795fac07e5046af04407454b235c012b9ef6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67270/
Test PASSed.

SparkQA commented:
**[Test build #67277 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67277/consoleFull)** for PR 15567 at commit [`80f26c6`](https://github.com/apache/spark/commit/80f26c6e010b18e712e9514146975f21af9ac8ab).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67277/
Test PASSed.

shearerp commented:
SPARK-14241 doesn't just occur with union and coalesce, it also occurs with filter and probably other operations. Hopefully this PR will address all of those situations. I strongly agree with the expected semantic in the original PR message by mengxr - this has bitten me on multiple occasions.

SparkQA commented:
**[Test build #67907 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67907/consoleFull)** for PR 15567 at commit [`9dcb249`](https://github.com/apache/spark/commit/9dcb249dd20d67ed8f3fa20e0fa0483322e3197a).

mengxr commented:
@rxin I updated the implementation to force initialization in Projection/Expression. This will fail many tests. I fixed all in `catalyst`, but not yet in `sql`. I want to propose the following change:

* add `Deterministic` extends `PartitionDependent` (always set initialized to true)
* remove `Nondeterministic`

Basically, we assume non-deterministic by default unless marked as deterministic. This will require updating all expressions but make the code less messy.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:
```diff
@@ -264,23 +264,43 @@ trait NonSQLExpression extends Expression {
+ * Trait for expressions that requires initialization based on the partition index prior to task
+ * execution on a partition.
+ */
+trait PartitionDependent {
```
as discussed offline this is StatefulExpression

mengxr replied:
Actually, `Projection` also need the same trait. Shall we call it `PartitionedStateful`?

SparkQA commented:
**[Test build #67907 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67907/consoleFull)** for PR 15567 at commit [`9dcb249`](https://github.com/apache/spark/commit/9dcb249dd20d67ed8f3fa20e0fa0483322e3197a).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67907/
Test FAILed.

mengxr committed:
Merge remote-tracking branch 'apache/master' into SPARK-14393.2

mengxr committed:
rename to initialize

SparkQA commented:
**[Test build #67957 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67957/consoleFull)** for PR 15567 at commit [`553c6a5`](https://github.com/apache/spark/commit/553c6a543dd18a7278bf989e9197e74dc3cece7c).

mengxr committed:
add doc

mengxr commented:
I reverted the changes I made to enforce `Projection.initialize`, which touched too many files and most of them doesn't really need to handle nondeterministic expressions. The current implementation at least throws a runtime error if a nondeterministic expression didn't get initialized, instead of returning incorrect result.

Renamed `initializeStatesForPartition` to `initialize`.

SparkQA commented:
**[Test build #67959 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67959/consoleFull)** for PR 15567 at commit [`ababaa9`](https://github.com/apache/spark/commit/ababaa9dfc7d38a7e276c61d5f35b28ad9718b19).

SparkQA commented:
**[Test build #67957 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67957/consoleFull)** for PR 15567 at commit [`553c6a5`](https://github.com/apache/spark/commit/553c6a543dd18a7278bf989e9197e74dc3cece7c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67957/
Test PASSed.

SparkQA commented:
**[Test build #67959 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67959/consoleFull)** for PR 15567 at commit [`ababaa9`](https://github.com/apache/spark/commit/ababaa9dfc7d38a7e276c61d5f35b28ad9718b19).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67959/
Test PASSed.

rxin commented:
Merging in master/branch-2.1. Thanks.

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala:
```diff
@@ -205,10 +206,11 @@ case class FilterExec(condition: Expression, child: SparkPlan)
-    child.execute().mapPartitionsInternal { iter =>
+    child.execute().mapPartitionsWithIndexInternal { (index, iter) =>
       val predicate = newPredicate(condition, child.output)
+      predicate.initialize(0)
```
Just wondering why `FilterExec` is not using `index` to initialize the conditions?
